core/src/main/scala/kafka/log/Log.scala

```diff
@@ -826,8 +832,16 @@ class Log(@volatile private var _dir: File,
         preallocate = config.preallocate))
     }
 
-    recoveryPoint = activeSegment.readNextOffset
```

Comment:
Can you help me understand what was wrong with this?

Reply:
Jun explained in the JIRA, The concern is that if there is a hard failure during recovery, you could end up with a situation where we persisted this, but we did not flush some of the segments. Does that make sense?

Reply:
Got it. So we may still be able to reopen an unflushed segment. That makes sense.

Reply:
Is it possible that a consumer could see "phantom" messages after recovery, even with this change?

1. Kafka Process dies with log data in page cache but not fsync'd
2. Recovery process sees the un-fsync'd log data but it looks ok so recovery succeeds, nothing to do.
3. Consumer fetches this data
4. OS hard dies, losing page cache
5. Consumer tries to repeat fetch from same offset but data has gone.

It seems to me once recovery has run we should be sure that all log segments are persistently stored. I'm not sure if we're currently providing that guarantee.

Reply:
@purplefox To clarify, the case you are talking about is:

1. A catastrophic scenario where a partition is offline (i.e. all replicas are down)
2. The OS was not shutdown
3. The OS did not flush the data to disk (this typically happens irrespective of our flushes due to OS configurations)
4. The replica that was the last member of the ISR comes back up, registers, to ZK and the Controller makes it the
leader (since it was the last member of the ISR, if it was a different replica, it won't be given leadership without
unclean leader election)
5. The hw is beyond the flushed segments
5. Consumer fetches the data beyond the flushed segments
6. OS hard dies

This is an interesting edge case, it seems incredibly unlikely, but possible if the hw can be beyond the flushed segments. @junrao & @hachikuji are we missing any detail that protects us against this?

The following has a similar impact:
1. Leader accepts a write
2. Write is replicated, hw is incremented, but data is not flushed
3. All replicas die, but the ISR is not shrunk yet
4. Leader receives a write, accepts it, replication doesn't happen since replicas are gone
5. ISR is shrunk, hw is incremented
6. Producer won't receive a successful ack given min.isr=2, but the consumer reads data that is only in the leader
7. Leader crashes and the unflushed data is gone (or hard disk dies and all the data in the leader is gone)

Flushing the segments during recovery helps on some scenarios, but not the one I just mentioned (assuming I am not missing anything). @hachikuji had a "strict min isr" proposal where the ISR is never allowed to shrink below `min.isr`. I haven't thought about all the details, but perhaps that covers both issues. Thoughts @hachikuji?

Reply:
For the case that Tim mentioned, if we defer advancing the recovery point, at step 5, the broker will be forced to do log recovery for all unflushed data. If the data is corrupted on disk, it will be detected during recovery. 

For the other case that Ismael mentioned, it is true that data can be lost in that case, but then this is the case where all replicas have failed.

Reply:
I think it is a gap that there is no minimum replication factor before a write can get exposed. Any writes that end up seeing the `NOT_ENOUGH_REPLICAS_AFTER_APPEND` error code are more vulnerable. These are unacknowledged writes, and the producer is expected to retry, but the consumer can still read them once the ISR shrinks and we would still view it as "data loss" if the broker failed before they could be flushed to disk. With the "strict min isr" proposal, the leader is not allowed to shrink the ISR lower than some replication factor, which helps to plug this hole.

Going back to @purplefox's suggestion, it does seem like a good idea to flush segments beyond the recovery point during recovery. It kind of serves to constrain the initial state of the system which makes it easier to reason about (e.g. you only need to worry about the loss of unflushed data from the previous restart). Some of the flush weaknesses probably still exist though regardless of this change.

Reply:
We discussed this offline and we decided to stick with the fix in this PR for now and to file a separate JIRA to consider flushing unflushed segments during recovery. That would provide stronger guarantees after a restart.

Reply:
Filed https://issues.apache.org/jira/browse/KAFKA-12386.